### PR TITLE
#99 - recipe category

### PR DESCRIPTION
**Pull request by hetec**
recipe category

- endpoints
- db access
- handlers
- tests

### CODE REVIEW

**Comment by flashback2k14** [comment]
bildschirmfoto 2017-12-25 um 11 27 38
could you please update your docker description with this PR!?

**Comment by flashback2k14** [comment]
bildschirmfoto 2017-12-25 um 11 29 51
i recommend to use the tslint plugin. it saves many time. :-)

File: server/src/recipe/handler/recipe.handler.ts [hunk 1]
```diff
@@ -19,19 +20,63 @@ export class RecipeHandler {
     };
   }
 
+  private _resolveCategory = (recipe: Recipe): Promise<Recipe> => {
```
Comment (flashback2k14) [inline server/src/recipe/handler/recipe.handler.ts]:
```ts
  private _resolveCategory = async (recipe: Recipe): Promise<Recipe> => {
    const result = await this._db.recipeCategories.getById(recipe.categoryId);
    recipe.category = result as RecipeCategory;
    return recipe;
  }
```
i would change it to use async await

File: server/src/recipe/handler/recipe.handler.ts [hunk 2]
```diff
@@ -19,19 +20,63 @@ export class RecipeHandler {
...
+    return this._db.recipeCategories.getById(recipe.categoryId)
+      .then((category: RecipeCategory) => {
+        recipe.category = category;
+        return recipe;
+      });
+  }
+
+  private _resolveAllCategories = (recipes: Recipe[]): Promise<Recipe[]> => {
```
Comment (flashback2k14) [inline server/src/recipe/handler/recipe.handler.ts]:
```ts
private _resolveAllCategories = async (recipes: Recipe[]): Promise<Recipe[]> => {
    return await Promise.all(recipes.map(async (recipe: Recipe) => {
      return await recipe.categoryId
        ? this._resolveCategory(recipe)
        : Promise.resolve(recipe);
    }));
  }
```
async await with map. looks simpler!?

File: server/src/recipe/handler/recipe.handler.ts [hunk 3]
```diff
@@ -19,19 +20,63 @@ export class RecipeHandler {
...
+    recipes.forEach((recipe: Recipe) => {
+      const recipeWithCategoryPromise = recipe.categoryId
+        ? this._resolveCategory(recipe)
+        : Promise.resolve(recipe);
+      promises = [...promises, recipeWithCategoryPromise];
+    });
+    return Promise.all(promises);
+  }
+
+  private _mapRecipesToRecipesWithCategory = (recipes: Recipe[]) => {
```
Comment (flashback2k14) [inline server/src/recipe/handler/recipe.handler.ts]:
```ts
private _mapRecipesToRecipesWithCategory = (recipes: Recipe[]) => {
    return recipes.map((recipe: Recipe) => {
      return new Recipe().clone(recipe).setCategoryId(undefined);
    });
  }
```

File: server/src/recipe/routes/recipe-category.route.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { ModelValidationHandler, ValidationGroup } from "../../app/index";
+import { AuthValidationHandler } from "../../auth/index";
+import { SocoboUserRoleType } from "../../socobouser/index";
+import { Recipe, RecipeCategoryHandler, RecipeMiddleware } from "../index";
+
+export class RecipeCategoryRoute {
+
+  constructor (
+    private _router: Router,
+    private _multer: Instance,
```
Comment (flashback2k14) [inline server/src/recipe/routes/recipe-category.route.ts]: i think this is not needed.

File: server/src/recipe/routes/recipe-category.route.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    private _recipeHandler: RecipeCategoryHandler,
+    private _authValidationHandler: AuthValidationHandler,
+    private _modelValidationHandler: ModelValidationHandler,
```
Comment (flashback2k14) [inline server/src/recipe/routes/recipe-category.route.ts]: no model validation is implemented.

File: server/src/recipe/routes/recipe-category.route.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    private _recipeMiddleware: RecipeMiddleware) {}
```
Comment (flashback2k14) [inline server/src/recipe/routes/recipe-category.route.ts]: i think this is not needed.

File: server/src/recipe/handler/recipe.handler.ts [hunk 7]
```diff
@@ -19,19 +20,63 @@ export class RecipeHandler {
...
+      } else {
+        res.status(200).json(result);
+      }
     } catch (error) {
       this._sendError(res)(error);
     }
   }
 
   public getAll = async (req: Request, res: Response) => {
+    const queryPramas = req.query;
```
Comment (flashback2k14) [inline server/src/recipe/handler/recipe.handler.ts]:
```ts
 public getAll = async (req: Request, res: Response) => {
    const queryPramas = req.query;
    try {
      const result = await this._db.recipe.getAll() as Recipe[];

      if (!queryPramas.hasOwnProperty("resolveCategory")) {
        res.status(200).json(result);
        return;
      }

      const recipesWithCategories = await this._resolveAllCategories(result);
      res.status(200).json(this._mapRecipesToRecipesWithCategory(recipesWithCategories));

    } catch (error) {
      this._sendError(res)(error);
    }
  }
```

**Comment by hetec** [comment]
Thanks for your feedback, I will incorporate your improvements

File: server/src/recipe/handler/recipe.handler.ts (hunk 2 above)
Comment (hetec) [reply server/src/recipe/handler/recipe.handler.ts]: Yes this looks better. But in this case you will wait for each _resolveCategor Promise to resolve. So the requests for each category are performed sequentially. The promise only approach will perform the requests in parallel. I will see that I can use the best of both, ok?

Comment (hetec) [reply server/src/recipe/handler/recipe.handler.ts]: With request I mean the db query for a category

**Pull request merged**